### ISSUE TITLE
Only show/count inteface references for public references in CodeLense

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#53559]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Only-showcount-inteface-references-for-/1417592)._

---
I'm looking at the CodeLense reference count in my interface file and I see 2 references.
One for the implementation and the second for my nameof() reference to the symbol.

This makes some sense in the implementation class when I look at the reference count (although I'm not sure it is useful as the reference count is intended). But it doesn't really seem to make sense in an interface file.
My usage of this reference count would ideally be just to show callers. And possibly implementers, but I can't see value in having the nameof() references. I suspect this feature is driven off of the "find all references" feature, which makes sense, but may not be the most useful.
**Should CodeLense filter out nameof() references in general? Or at least for interface files?**

Note: I absolutely see the benefit of nameof() for renaming symbols, which is why I use nameof(). But am challenging the benefit in those uses being surfaced for human analysis via CodeLense.

! [image.png] (https://aka.ms/dc/image?name=B650d483d39b745358d117fd8c594a753637562587844051361_20210510-105304-image.png&tid=650d483d39b745358d117fd8c594a753637562587844051361)

---
#### Feedback Bot on 5/10/2021, 11:02 PM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by sharwell** [comment]
References to an item by name, including `nameof` and `cref` (in documentation), are considered references for the purpose of Find All References and CodeLens. I'm not sure we can make a change to this behavior without significantly impairing users with workflows based on the current behavior.

**Comment by CyrusNajmabadi** [comment]
I'm also not really sure why this would be valuable for 'find all references', but not for 'code lens references'.

**Comment by sbrockway** [comment]
Agreed, is it even valuable to have nameof() and cref come back with the find all references? It certainly is for F2 renaming.

**Comment by CyrusNajmabadi** [comment]
> Agreed, is it even valuable to have nameof() and cref come back with the find all references?

Yes...  Why would they not be?  If I ask to get an the references, and I see none, and I remove the member, and now the code doesn't compile because of the nameof reference, that seems... Bad :-)

**Comment by sharwell** [comment]
@sbrockway it appears you might be the original author of this request. While it's extremely unlikely we would make this change _as described_, I'm interested to know more about the scenario where you felt this would be valuable. Can you describe the situation you were facing, the way(s) you were interacting with CodeLens, and details about how it would have saved you time or effort if these elements were omitted? Perhaps there is another way we can work on the design to help you achieve the same outcome.

**Comment by CyrusNajmabadi** [comment]
> When I'm looking at an interface member, why would I need to see CodeLens information for nameof() references? 

Because its' a reference to that member :)  If i want to see how that member is used, a `nameof` usage would be very relevant.  It's a place where some other piece of code is taking a dependency on the interface member...

I think we have the conversation reversed here.  I think i'd need a good explanation why these valid reference dependencies would *not* be shown.  THey seem relevant and important, and i would certainly want to know (and would not want CodeLens to tell me there were zero references when in fact there was a dep like this). :)

**Comment by sbrockway** [comment]
> 
> 
> > Agreed, is it even valuable to have nameof() and cref come back with the find all references?
> 
> Yes... Why would they not be? If I ask to get an the references, and I see none, and I remove the member, and now the code doesn't compile because of the nameof reference, that seems... Bad :-)

Would you decide to keep a member solely base on the nameof() or cref references? Or would you make that decision base on the callers present in your code? For me its the later. If the former result in compilation errors, then good. I know what to cleanup after I remove the member. If I have no callers left, but CodeLense shows me "1 reference" because I have a nameof() in my implementation that doesn't seem useful to me.

**Comment by sharwell** [comment]
@sbrockway If I understand the request correctly, you are looking for an improved experience to locate interface members for which there are no known callers. I would add that it could be beneficial to further have the following be clear:

* Callers that directly reference an implementation (i.e. the interface method could go away as long as the implementation stays)
* Callers that occur only in test code (e.g. an interface method that is never used in production but sticks around because it's still getting tested)

**Comment by sbrockway** [comment]
@sharwell Agreed.

**Comment by sharwell** [comment]
@mikadumont can you make sure https://github.com/dotnet/roslyn/issues/53559#issuecomment-846114395 is tracked with the open navigation and search stories?

**Comment by vatsalyaagrawal** [comment]
Design meeting notes:

- Could potentially be used to delete code. This is a hard problem so solve. This information could potentially be surfaced in CodeLens or someplace else.
- If deleting code is user intent, we should have separate functionality to address this scenario.  
- Can we provide more information in CodeLens for user to take correct action?
- Could potentially add columns similar to read/write in find all references. 
- We need to understand user scenarios. What will users use this data for? 

Decision: Needs proposal to consolidate all options/thoughts/suggestions.
